File: TSL UHF ASCII 2 SDK DotNet v1.1/ASCII Protocol Inventory Sample/AsciiProtocolInventoryFx/Views/MainForm.cs
//-----------------------------------------------------------------------
// <copyright file="MainForm.cs" company="Technology Solutions UK Ltd"> 
//     Copyright (c) 2014 Technology Solutions UK Ltd. All rights reserved. 
// </copyright> 
// <author>Robin Stone</author>
//-----------------------------------------------------------------------
namespace AsciiProtocolInventory.Views
{
    using System;
    using System.Collections.Generic;
    using System.ComponentModel;
    using System.Data;
    using System.Drawing;
    using System.Linq;
    using System.Text;
    using System.Windows.Forms;
    
    using TechnologySolutions.Rfid.AsciiProtocol;
    using ViewModels;

    public partial class MainForm : Form
    {
        /// <summary>
        /// Handles conencting to the reader
        /// </summary>
        private ConnectViewModel connectViewModel;

        /// <summary>
        /// Handles commanding the reader
        /// </summary>
        private MainViewModel viewModel;

        /// <summary>
        /// Initializes a new instance of the MainForm class
        /// </summary>
        public MainForm()
        {
            this.InitializeComponent();

            this.viewModel = Service.Instance.MainViewModel;
            this.viewModel.TransponderMessage += this.ViewModel_TransponderMessage;
            this.viewModel.BarcodeMessage += this.ViewModel_BarcodeMessage;
            this.viewModel.ResponseLine += this.ViewModel_ResponseLine;

            // TODO: this.viewModel.PropertyChanged 
            this.connectViewModel = Service.Instance.ConnectViewModel;
            this.connectViewModel.PropertyChanged += this.ConnectViewModel_PropertyChanged;

            // Create a menu of session values
            this.querySessionComboBox.DataSource = Enum.GetValues(typeof(QuerySession));

            // Use the description when displaying to the user
            this.querySessionComboBox.FormattingEnabled = true;
            this.querySessionComboBox.Format += delegate(object sender, ListControlConvertEventArgs e)
            {
                e.Value = ((QuerySession)e.Value).Description();
            };

            this.queryTargetComboBox.DataSource = Enum.GetValues(typeof(QueryTarget));
            this.queryTargetComboBox.FormattingEnabled = true;
            this.queryTargetComboBox.Format += delegate(object sender, ListControlConvertEventArgs e)
            {
                e.Value = ((QueryTarget)e.Value).Description();
            };
            
            this.Load += this.Form_Load;
        }

        /// <summary>
        /// Gets or sets the list of available com ports
        /// </summary>
        public IEnumerable<string> PortNames
        {
            get
            {
                foreach (ToolStripDropDownItem item in this.portToolStripMenuItem.DropDownItems)
                {
                    yield return item.Name;
                }
            }

            set
            {
                string selectedPortName;

                selectedPortName = this.connectViewModel.PortName;
                this.portToolStripMenuItem.DropDownItems.Clear();
                foreach (string portName in value)
                {
                    ToolStripMenuItem menuItem;

                    menuItem = new ToolStripMenuItem()
                    {
                        Text = portName,
                        Checked = selectedPortName.Equals(portName)
                    };

                    menuItem.Click += delegate(object sender, EventArgs e)
                    {
                        foreach (ToolStripMenuItem item in this.portToolStripMenuItem.DropDownItems)
                        {
                            item.Checked = false;
                        }

                        this.connectViewModel.PortName = (sender as ToolStripMenuItem).Text;
                        (sender as ToolStripMenuItem).Checked = true;
                    };

                    this.portToolStripMenuItem.DropDownItems.Add(menuItem);
                }
            }
        }

        private void Form_Load(object sender, EventArgs e)
        {
            this.PortNames = this.connectViewModel.PortNames;

            // Configure minimum sizes here to work round a UI designer bug
            this.resultsSplitContainer.Panel1MinSize = 400;
            this.resultsSplitContainer.Panel2MinSize = 300;
            this.mainSplitContainer.Panel1MinSize = 120;
            this.mainSplitContainer.Panel2MinSize = 240;

            this.mainSplitContainer.Panel1Collapsed = !this.viewModel.IsProtocolResponseVisible;
            this.showProtocolResponsesToolStripMenuItem.Checked = this.viewModel.IsProtocolResponseVisible;

            this.ConnectViewModel_PropertyChanged(sender, new PropertyChangedEventArgs("IsConnected"));
        }

        private void ExitToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.Close();
        }

        private void ConnectToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.connectViewModel.Connect();
        }

        private void DisconnectToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.connectViewModel.Disconnect();
        }

        private void RefreshPortsToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.connectViewModel.RefreshPorts();
            this.PortNames = this.connectViewModel.PortNames;
        }

        private void ScanRfidCommandButton_Click(object sender, EventArgs e)
        {
            System.Threading.ThreadPool.QueueUserWorkItem(delegate(object state)
            {
                // don't block the UI while the command executes
                this.viewModel.ExecuteInventoryCommandSynchronously();
            });
        }

        private void scanRfidAsyncButton_Click(object sender, EventArgs e)
        {
            System.Threading.ThreadPool.QueueUserWorkItem(delegate(object state)
            {
                // don't block the UI while the command executes
                this.viewModel.ExecuteInventoryCommandAsynchronously();
            });
        }       

        private void ScanBarcodeCommandButton_Click(object sender, EventArgs e)
        {
            System.Threading.ThreadPool.QueueUserWorkItem(delegate(object state)
            {
                // don't block the UI while the command executes
                this.viewModel.ExecuteBarcodeCommandSynchronously();
            });
        }

        private void ScanBarcodeAsyncButton_Click(object sender, EventArgs e)
        {
            System.Threading.ThreadPool.QueueUserWorkItem(delegate(object state)
            {
                // don't block the UI while the command executes
                this.viewModel.ExecuteBarcodeCommandAsynchronously();
            });
        }

        private void AbortBarcodeButton_Click(object sender, EventArgs e)
        {
            System.Threading.ThreadPool.QueueUserWorkItem(delegate(object state)
            {
                this.viewModel.ExecuteAbortCommand();
            });
        }

        private void ViewModel_TransponderMessage(object sender, TextEventArgs e)
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new EventHandler<TextEventArgs>(this.ViewModel_TransponderMessage), new object[] { sender, e });
            }
            else
            {
                this.DisplayText(this.transponderListBox, e.Text);
            }
        }

        private void ViewModel_BarcodeMessage(object sender, TextEventArgs e)
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new EventHandler<TextEventArgs>(this.ViewModel_BarcodeMessage), new object[] { sender, e });
            }
            else
            {
                this.DisplayText(this.barcodeListBox, e.Text);
            }
        }

        private void ViewModel_ResponseLine(object sender, TextEventArgs e)
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new EventHandler<TextEventArgs>(this.ViewModel_ResponseLine), new object[] { sender, e });
            }
            else
            {
                this.DisplayText(this.responsesListBox, e.Text);
            }
        }

        private void ConnectViewModel_PropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new PropertyChangedEventHandler(this.ConnectViewModel_PropertyChanged), new object[] { sender, e });
            }
            else
            {
                this.connectToolStripMenuItem.Enabled = this.connectViewModel.CanConnect();
                this.disconnectToolStripMenuItem.Enabled = this.connectViewModel.CanDisconnect();
                this.readerToolStripMenuItem.Enabled = this.connectViewModel.CanRefreshPorts();
                this.messageToolStripStatusLabel.Text = this.connectViewModel.ConnectionStatus;
            }
        }

        private void DisplayText(ListBox listBox, string value)
        {
            listBox.SuspendLayout();
            if (listBox.Items.Count > 10000)
            {
                for (int i = 0; i < 50; i++)
                {
                    listBox.Items.RemoveAt(0);
                }
            }

            listBox.Items.Add(value);
            listBox.SelectedIndex = listBox.Items.Count - 1;
            listBox.ResumeLayout();
        }

        private void AboutToolStripMenuItem_Click(object sender, EventArgs e)
        {
            using (Form about = new AboutBoxForm())
            {
                about.ShowDialog();
            }
        }

        private void ClearTransponderResponsesButton_Click(object sender, EventArgs e)
        {
            this.transponderListBox.Items.Clear();
        }

        private void ClearBarcodeResponsesButton_Click(object sender, EventArgs e)
        {
            this.barcodeListBox.Items.Clear();
        }

        private void ClearProtocolResponsesToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.responsesListBox.Items.Clear();
        }

        private void ClearTransponderResponsesToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.transponderListBox.Items.Clear();
        }

        private void ClearBarcodeResponsesToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.barcodeListBox.Items.Clear();
        }

        private void ClearAllResponsesToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.responsesListBox.Items.Clear();
            this.transponderListBox.Items.Clear();
            this.barcodeListBox.Items.Clear();
        }

        private void ShowProtocolResponsesToolStripMenuItem_Click(object sender, EventArgs e)
        {
            bool newState = !this.viewModel.IsProtocolResponseVisible;
            this.mainSplitContainer.Panel1Collapsed = !newState;
            this.showProtocolResponsesToolStripMenuItem.Checked = newState;
            this.viewModel.IsProtocolResponseVisible = newState;
        }

        private void QuerySessionComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            this.viewModel.Session = (QuerySession)querySessionComboBox.SelectedItem;
        }

        private void QueryTargetComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            this.viewModel.Target = (QueryTarget)queryTargetComboBox.SelectedItem;
        }        
    }
}

File: TSL UHF ASCII 2 SDK DotNet v1.1/ASCII Protocol Inventory Sample VB/AsciiProtocolInventoryFx.ConvertedToVBNet/Views/MainForm.vb
Imports TechnologySolutions.Rfid.AsciiProtocol
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Namespace Views

	Public Partial Class MainForm
		Inherits Form
		''' <summary>
		''' Handles conencting to the reader
		''' </summary>
		Private connectViewModel As ConnectViewModel

		''' <summary>
		''' Handles commanding the reader
		''' </summary>
		Private viewModel As MainViewModel

		''' <summary>
		''' Initializes a new instance of the MainForm class
		''' </summary>
		Public Sub New()
			Me.InitializeComponent()

			Me.viewModel = New MainViewModel()
			AddHandler Me.viewModel.CommandCompleted, AddressOf Me.ViewModel_CommandCompleted
			AddHandler Me.viewModel.TransponderMessageHandler, AddressOf Me.ViewModel_TransponderMessage
			AddHandler Me.viewModel.BarcodeMessageHandler, AddressOf Me.ViewModel_BarcodeMessage
			AddHandler Me.viewModel.ResponseLine, AddressOf Me.ViewModel_ResponseLine
			' TODO: this.viewModel.PropertyChanged 

			Me.connectViewModel = New ConnectViewModel()
			Me.connectViewModel.PortName = Service.Settings.PortName
			AddHandler Me.connectViewModel.PropertyChanged, AddressOf Me.ConnectViewModel_PropertyChanged

			' Create a menu of session values
			Me.querySessionComboBox.DataSource = [Enum].GetValues(GetType(QuerySession))
			' Use the description when displaying to the user
            Me.querySessionComboBox.FormattingEnabled = True
            AddHandler Me.querySessionComboBox.Format, AddressOf Me.ComboBox_EnumFormat

            Me.queryTargetComboBox.DataSource = [Enum].GetValues(GetType(QueryTarget))
            Me.queryTargetComboBox.FormattingEnabled = True
            AddHandler Me.queryTargetComboBox.Format, AddressOf Me.ComboBox_EnumFormat

            AddHandler Me.Load, AddressOf Me.Form_Load
        End Sub

        ''' <summary>
        ''' Gets or sets the list of available com ports
        ''' </summary>
        Public Property PortNames() As IEnumerable(Of String)
            Get
                Dim result As List(Of String)

                result = New List(Of String)
                For Each item As ToolStripDropDownItem In Me.portToolStripMenuItem.DropDownItems
                    result.Add(item.Name)
                Next

                Return result
            End Get

            Set(ByVal value As IEnumerable(Of String))
                Dim selectedPortName As String

                selectedPortName = Me.connectViewModel.PortName
                Me.portToolStripMenuItem.DropDownItems.Clear()
                For Each portName As String In value
                    Dim menuItem As ToolStripMenuItem

                    menuItem = New ToolStripMenuItem() With { _
                      .Text = portName, _
                      .Checked = selectedPortName.Equals(portName) _
                    }

                    AddHandler menuItem.Click, AddressOf PortMenuItem_Click

                    Me.portToolStripMenuItem.DropDownItems.Add(menuItem)
                Next
            End Set
        End Property

        Private Sub PortMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            For Each item As ToolStripMenuItem In Me.portToolStripMenuItem.DropDownItems
                item.Checked = False
            Next

            Me.connectViewModel.PortName = TryCast(sender, ToolStripMenuItem).Text
            TryCast(sender, ToolStripMenuItem).Checked = True
        End Sub

        ' returns the description of the Enum to display in the combobox
        Private Sub ComboBox_EnumFormat(ByVal sender As Object, ByVal e As ListControlConvertEventArgs)
            If e.Value Is GetType(QuerySession) Then
                e.Value = CType(e.Value, QuerySession).Description()
            ElseIf e.Value Is GetType(QueryTarget) Then
                e.Value = CType(e.Value, QueryTarget).Description()
            End If
        End Sub

        Private Sub Form_Load(ByVal sender As Object, ByVal e As EventArgs)
            Me.PortNames = Me.connectViewModel.PortNames

            ' Configure minimum sizes here to work round a UI designer bug
            Me.resultsSplitContainer.Panel1MinSize = 400
            Me.resultsSplitContainer.Panel2MinSize = 300
            Me.mainSplitContainer.Panel1MinSize = 120
            Me.mainSplitContainer.Panel2MinSize = 240


            Me.mainSplitContainer.Panel1Collapsed = Not Service.Settings.isProtocolResponsePanelVisible
            Me.showProtocolResponsesToolStripMenuItem.Checked = Service.Settings.isProtocolResponsePanelVisible

            Me.ConnectViewModel_PropertyChanged(sender, New PropertyChangedEventArgs("IsConnected"))
        End Sub

        Private Sub ExitToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.Close()
        End Sub

        Private Sub ConnectToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.connectViewModel.Connect()
        End Sub

        Private Sub DisconnectToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.connectViewModel.Disconnect()
        End Sub

        Private Sub RefreshPortsToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.connectViewModel.RefreshPorts()
            Me.PortNames = Me.connectViewModel.PortNames
        End Sub

        Private Sub ScanRfidCommandButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            System.Threading.ThreadPool.QueueUserWorkItem(AddressOf Me.ExecuteInventoryAsync)
        End Sub

        Private Sub scanRfidAsyncButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.viewModel.ExecuteInventoryCommand(False)
        End Sub

        ' This will be called on the thread pool thread. 
        ' As executing synchronously the command will block until the response is received.
        ' Calling from a thread pool thread leaves the UI thread free to update as the response is received
        Private Sub ExecuteInventoryAsync(ByVal state As Object)
            Me.viewModel.ExecuteInventoryCommand(True)
        End Sub

        Private Sub ScanBarcodeCommandButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            System.Threading.ThreadPool.QueueUserWorkItem(AddressOf Me.ExecuteBarcodeAsync)
        End Sub

        Private Sub ScanBarcodeAsyncButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.viewModel.ExecuteBarcodeCommand(False)
        End Sub

        ' This will be called on the thread pool thread. 
        ' As executing synchronously the command will block until the response is received.
        ' Calling from a thread pool thread leaves the UI thread free to update as the response is received
        Private Sub ExecuteBarcodeAsync(ByVal state As Object)
            Me.viewModel.ExecuteBarcodeCommand(True)
        End Sub

        Private Sub ViewModel_CommandCompleted(ByVal sender As Object, ByVal e As EventArgs)
            If Me.InvokeRequired Then
                Me.Invoke(New EventHandler(AddressOf Me.ViewModel_CommandCompleted), New Object() {sender, e})
            Else
            End If
        End Sub

        Private Sub ViewModel_TransponderMessage(ByVal sender As Object, ByVal e As TextEventArgs)
            If Me.InvokeRequired Then
                Me.Invoke(New EventHandler(Of TextEventArgs)(AddressOf Me.ViewModel_TransponderMessage), New Object() {sender, e})
            Else
                Me.DisplayText(Me.transponderListBox, e.Text)
            End If
        End Sub

        Private Sub ViewModel_BarcodeMessage(ByVal sender As Object, ByVal e As TextEventArgs)
            If Me.InvokeRequired Then
                Me.Invoke(New EventHandler(Of TextEventArgs)(AddressOf Me.ViewModel_BarcodeMessage), New Object() {sender, e})
            Else
                Me.DisplayText(Me.barcodeListBox, e.Text)
            End If
        End Sub

        Private Sub ViewModel_ResponseLine(ByVal sender As Object, ByVal e As TextEventArgs)
            If Me.InvokeRequired Then
                Me.Invoke(New EventHandler(Of TextEventArgs)(AddressOf Me.ViewModel_ResponseLine), New Object() {sender, e})
            Else
                Me.DisplayText(Me.responsesListBox, e.Text)
            End If
        End Sub

        Private Sub ConnectViewModel_PropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
            If Me.InvokeRequired Then
                Me.Invoke(New PropertyChangedEventHandler(AddressOf Me.ConnectViewModel_PropertyChanged), New Object() {sender, e})
            Else
                Me.connectToolStripMenuItem.Enabled = Me.connectViewModel.CanConnect()
                Me.disconnectToolStripMenuItem.Enabled = Me.connectViewModel.CanDisconnect()
                Me.readerToolStripMenuItem.Enabled = Me.connectViewModel.CanRefreshPorts()
                Me.messageToolStripStatusLabel.Text = Me.connectViewModel.ConnectionStatus
            End If
        End Sub

        Private Sub DisplayText(ByVal listBox As ListBox, ByVal value As String)
            listBox.SuspendLayout()
            If listBox.Items.Count > 10000 Then
                For i As Integer = 0 To 49
                    listBox.Items.RemoveAt(0)
                Next
            End If

            listBox.Items.Add(value)
            listBox.SelectedIndex = listBox.Items.Count - 1
            listBox.ResumeLayout()
        End Sub


        Private Sub AboutToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Using about As Form = New AboutBoxForm()
                about.ShowDialog()
            End Using
        End Sub

        Private Sub ClearTransponderResponsesButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.transponderListBox.Items.Clear()
        End Sub

        Private Sub ClearBarcodeResponsesButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.barcodeListBox.Items.Clear()
        End Sub

        Private Sub ClearProtocolResponsesToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.responsesListBox.Items.Clear()
        End Sub

        Private Sub ClearTransponderResponsesToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.transponderListBox.Items.Clear()
        End Sub

        Private Sub ClearBarcodeResponsesToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.barcodeListBox.Items.Clear()
        End Sub

        Private Sub ClearAllResponsesToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.responsesListBox.Items.Clear()
            Me.transponderListBox.Items.Clear()
            Me.barcodeListBox.Items.Clear()
        End Sub

        Private Sub ShowProtocolResponsesToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim newState As Boolean = Not Service.Settings.isProtocolResponsePanelVisible
            Me.mainSplitContainer.Panel1Collapsed = Not newState
            Me.showProtocolResponsesToolStripMenuItem.Checked = newState
            Service.Settings.isProtocolResponsePanelVisible = newState
        End Sub

        Private Sub MainForm_FormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs)
            Service.Settings.PortName = Me.connectViewModel.PortName
            Service.Settings.Save()
        End Sub

        Private Sub QuerySessionComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            Me.viewModel.Session = CType(querySessionComboBox.SelectedItem, QuerySession)
        End Sub

        Private Sub QueryTargetComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            Me.viewModel.Target = CType(queryTargetComboBox.SelectedItem, QueryTarget)
        End Sub
    End Class
End Namespace
